javascript/ql/src/experimental/Security/CWE-079/ClipboardXss.ql

```diff
@@ -0,0 +1,78 @@
+/**
+ * @name Client-side clipboard-based cross-site scripting
+ * @description Pasting clipboard input directly to the DOM without proper sanitization allows for
+ *              a cross-site scripting vulnerability.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 3.1
+ * @precision high
+ * @id js/clipboard-xss
+ * @tags security
+ *       external/cwe/cwe-079
+ */
+
+import javascript
+import DataFlow
+import semmle.javascript.security.dataflow.DomBasedXssQuery as DomBasedXss
+import DataFlow::PathGraph
+
+/*
+ * Gets references to clipboardData DataTransfer objects
+ * https://developer.mozilla.org/en-US/docs/Web/API/ClipboardEvent/clipboardData
+ */
+
+SourceNode clipboardDataTransferSource(TypeTracker t) {
+  t.start() and
+  exists(DataFlow::PropRead pr | pr.getPropertyName() = "clipboardData" and pr.flowsTo(result))
+  or
+  exists(TypeTracker t2 | result = clipboardDataSource(t2).track(t2, t))
+}
+
+SourceNode clipboardDataTransferSource() {
+  result = clipboardDataTransferSource(TypeTracker::end())
+}
+
+/*
+ * Gets references to the result of a call to getData on a DataTransfer object
+ * https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/getData
+ */
+
+SourceNode clipboardDataSource(TypeTracker t) {
+  t.start() and
+  clipboardDataTransferSource().getAMethodCall("getData").flowsTo(result)
+  or
+  exists(TypeTracker t2 | result = clipboardDataSource(t2).track(t2, t))
+}
+
+SourceNode clipboardDataSource() { result = clipboardDataSource(TypeTracker::end()) }
+
+class ClipboardSource extends DataFlow::Node {
+  ClipboardSource() { this = clipboardDataSource() }
```

Comment:
Just a minor comment here, not something that needs to be fixed in this PR as we're planning to move the source into another query anyway.

Type tracking the source forward can lead to duplicate alerts (and worse performance), as each node you reach becomes a new source, but the original will also reach the sink. For example, here we would get two alerts, one with the original source (`getData()`) and one with `x` as the source.
```js
let source1 = transfer.getData(); // alert from here
f(source1);

function f(x) {  // also alert from here
  $(x); // sink
}
```
The solution would be to just use `clipboardDataTransferSource().getAMethodCall("getData")` as the only source.

Reply:
Sure, this was probably a result of me trying to fix the before mentioned typo. I can confirm using ```clipboardDataTransferSource().getAMethodCall("getData")``` as the only source is able to detect both of the vulnerabilities mentioned in #6498.

Reply:
@asgerf can you propose this change as well?

Thanks!